### DISCUSSION TITLE
xpack 制作nsni安装包时如何指定安装完成之后执行的脚本，我尝试了xpack:after_installcmd，但是会在打包时运行

### DISCUSSION THREAD

**Discussion by cpcgskill** [xmake-io/xmake#5002] [Q&A]
### 背景
xpack 在制作nsni安装包时如何指定安装完成之后执行的脚本，我尝试了xpack:after_installcmd，但是会在打包时，而不是安装后运行。

### 猜测  
我看到这样一句话：
```
需要注意的是，通过 batchcmds 添加的 cp, mkdir 等命令都不会被立即执行，而是仅仅生成一个命令列表，后面实际生成包的时候，会将这些命令，翻译成打包命令。
```
我猜测这一段话的意思是不是，这代码会在打包时执行，然后记录下执行的命令之后，在安装时执行。

### 打包代码
如果不是我猜测的这样，麻烦大佬们看看我的打包设置是不是有问题，打包代码如下：
```lua
includes("@builtin/xpack")
xpack("AModFile")
    -- 一部分设置
    after_installcmd(function (package, batchcmds)
        user_directory = os.getenv("USERPROFILE")
        documents_directory = user_directory .. "\\Documents"
        maya_directory = documents_directory .. "\\maya"
        modules_directory = maya_directory .. "\\modules"
        mod_file = modules_directory .. "\\AModFile.mod"
        mod_string = ""
        for i, v in ipairs(version_list) do
            print("Write to " .. v)
            mod_string = mod_string .. "+ PLATFORM:win64 MAYAVERSION:" .. v .. " AModFile" .. package:version_str() .. " " .. package:installdir() .. "\n"
            mod_string = mod_string .. "[r] plug-ins: .\n"
            mod_string = mod_string .. "[r] scripts: .\n"
        end
        if not os.isdir(modules_directory) then
            os.mkdir(modules_directory)
        end
        print("Write to " .. mod_file)
        print(mod_string)
        io.writefile(mod_file, mod_string)
    end)
```

**Comment by waruqi** [comment]
安装命令，得用 batchcmds:cp ... 而不是 os.cp

**Reply by cpcgskill** [reply]
我目前将这部分逻辑放到了`before_package`， 现在`after_installcmd`里执行的是`batchcmds:cp(package:installdir("mod/*.mod"), "$(USERPROFILE)/Documents/maya/modules")`。
  但是发现还是不行，我猜测是因为这个是packages:instaldir()拿到的不是安装的目标目录的原因，但是我不清楚该如何获得安装的目标目录。

  我目前希望的是执行用户执行安装程序时将一个文件拷贝到一个找到目录下（因为现在我写的不是一个桌面软件，是另一个软件的插件，得拷贝一个配置过去），有什么好办法吗？

**Reply by waruqi** [reply]
不可能拿不到，xmake 自己的打包里面都是这么用的 https://github.com/xmake-io/xmake/blob/316c9cdee35d47e57219412d7ad39326966626c0/core/xpack.lua#L124

https://github.com/xmake-io/xmake/blob/316c9cdee35d47e57219412d7ad39326966626c0/xmake/plugins/pack/xpack.lua#L398

https://github.com/xmake-io/xmake/blob/316c9cdee35d47e57219412d7ad39326966626c0/tests/plugins/pack/xmake.lua#L49

`$(USERPROFILE)` 这种不支持

**Reply by waruqi** [reply]
另外，package:installdir 是目的安装目录，不是源目录，还有安装包最好只安装文件到自己的 installdir 内部，不要安装到外面

**Reply by cpcgskill** [reply]
感谢，不过我仍然不太理解，可能得在研究一下。
我了解了一下nsis，发现它使用了一种独立的脚本语言，我根据nsis的脚本尝试了`batchcmds:cp("mod/*.mod", "$DOCUMENTS/maya/modules")`之后似乎仍然是错误的。
但是我还尝试了
```lua
batchcmds:rawcmd("nsis", [[
SetShellVarContext current
CopyFiles "$INSTDIR\mod\DQMorphSkin.mod" "$DOCUMENTS\maya\modules"
]])
```
这个是能够正常运行的。
我想知道是否有文档讲解了，这部分xmake.lua脚本与nsis脚本之间的关系。

**Reply by cpcgskill** [reply]
`另外，package:installdir 是目的安装目录，不是源目录，还有安装包最好只安装文件到自己的 installdir 内部，不要安装到外面`
是的确实如此，需要拷贝的文件仅有一个，是用于让插件所属的软件发现插件的所在的。

**Reply by waruqi** [reply]
> 我根据nsis的脚本尝试了batchcmds:cp("mod/*.mod", "$DOCUMENTS/maya/modules")之后似乎仍然是错误的。

刚说了 `$DOCUMENTS` 不支持，除非你走 rawcmd 否则不要使用 nsis 的变量

**Reply by cpcgskill** [reply]
我明白了（大概），目前使用了rawcmd解决了问题，batchcmds部分我后面慢研究🤣
